File: 002_Systematic_Trading/002-Chan_Jegadeesh_Lakonishok-Momentum_Strategies-1996/code/object_oriented/cpp_23/src/momentum_calculator.cpp
#include <algorithm>
#include <cmath>
#include <fstream>
#include <iostream>
#include <ranges>
#include <sstream>
#include <string>
#include <vector>
#include <flat_map>
#include <format>

struct EpsData
{
    std::string company_id;
    std::string quarter;
    double eps;

    static EpsData from_csv_line(const std::string &line)
    {
        std::stringstream ss(line);
        std::string company_id, quarter, eps_str;
        std::getline(ss, company_id, ',');
        std::getline(ss, quarter, ',');
        std::getline(ss, eps_str, ',');
        return {company_id, quarter, std::stod(eps_str)};
    }
};

class MomentumCalculator
{
public:
    explicit MomentumCalculator(int lookback = 4) : lookback_(lookback) {}

    std::vector<std::pair<std::string, double>> compute_signal(
        const std::vector<EpsData> &eps_data) const
    {
        // Group by company
        std::flat_map<std::string, std::vector<EpsData>> company_eps;
        for (const auto &record : eps_data)
        {
            company_eps[record.company_id].push_back(record);
        }

        std::vector<std::pair<std::string, double>> results;
        for (auto &[company_id, records] : company_eps)
        {
            // Sort by quarter
            std::ranges::sort(records, {}, &EpsData::quarter);
            if (records.size() < static_cast<size_t>(lookback_ + 1))
            {
                continue;
            }

            // Compute momentum
            const double latest_eps = records.back().eps;
            const double past_eps = records[records.size() - 1 - lookback_].eps;
            std::vector<double> eps_values;
            for (const auto &r : records | std::views::take_last(5))
            {
                eps_values.push_back(r.eps);
            }
            const double sigma = stddev(eps_values);
            if (sigma == 0.0)
            {
                continue;
            }
            results.emplace_back(company_id, (latest_eps - past_eps) / sigma);
        }
        return results;
    }

private:
    int lookback_;

    static double stddev(const std::vector<double> &values)
    {
        if (values.empty())
            return 0.0;
        const double mean = std::ranges::sum(values) / values.size();
        const double variance = std::ranges::fold_left(
                                    values | std::views::transform([&](double x)
                                                                   { return (x - mean) * (x - mean); }),
                                    0.0,
                                    std::plus{}) /
                                values.size();
        return std::sqrt(variance);
    }
};

std::vector<EpsData> load_eps_data(const std::string &file_path)
{
    std::ifstream file(file_path);
    std::vector<EpsData> data;
    std::string line;
    std::getline(file, line); // Skip header
    while (std::getline(file, line))
    {
        data.push_back(EpsData::from_csv_line(line));
    }
    return data;
}

int main()
{
    MomentumCalculator calc;
    auto eps_data = load_eps_data("eps_data.csv");
    auto signals = calc.compute_signal(eps_data);
    for (const auto &[company_id, signal] : signals)
    {
        std::cout << std::format("{}: {:.4f}\n", company_id, signal);
    }
    return 0;
}

File: 002_Systematic_Trading/002-Chan_Jegadeesh_Lakonishok-Momentum_Strategies-1996/code/functional/python_3_11/src/momentum_calculator.py
from typing import List, Tuple, Dict
from functools import reduce
import csv
import math
from collections import defaultdict

def parse_quarter(quarter: str) -> Tuple[int, int]:
    match quarter.split('_'):
        case ['Q1', year]:
            return (int(year), 1)
        case ['Q2', year]:
            return (int(year), 2)
        case ['Q3', year]:
            return (int(year), 3)
        case ['Q4', year]:
            return (int(year), 4)
        case _:
            raise ValueError(f"Invalid quarter format: {quarter}")

def stddev(values: List[float]) -> float:
    if not values:
        return 0.0
    mean = sum(values) / len(values)
    variance = sum((x - mean) ** 2 for x in values) / len(values)
    return math.sqrt(variance)

def compute_momentum_signal(
    eps_data: List[Dict[str, str]], lookback: int = 4
) -> List[Tuple[str, float]]:
    # Group by company
    company_eps: Dict[str, List[Dict[str, str]]] = defaultdict(list)
    for record in eps_data:
        company_eps[record['company_id']].append(record)

    def process_company(
        company_id: str, records: List[Dict[str, str]]
    ) -> Tuple[str, float] | None:
        # Sort by quarter
        sorted_records = sorted(records, key=lambda x: parse_quarter(x['quarter']))
        if len(sorted_records) < lookback + 1:
            return None

        # Compute momentum
        latest_eps = float(sorted_records[-1]['eps'])
        past_eps = float(sorted_records[-1 - lookback]['eps'])
        eps_values = [float(r['eps']) for r in sorted_records[-5:]]
        sigma = stddev(eps_values)
        if sigma == 0:
            return None
        return (company_id, (latest_eps - past_eps) / sigma)

    # Map and filter valid results
    results = [
        result
        for result in map(
            lambda x: process_company(x[0], x[1]), company_eps.items()
        )
        if result is not None
    ]

    return results

def load_eps_data(file_path: str) -> List[Dict[str, str]]:
    with open(file_path, 'r') as f:
        return list(csv.DictReader(f))

if __name__ == "__main__":
    eps_data = load_eps_data("eps_data.csv")
    signals = compute_momentum_signal(eps_data)
    for company_id, signal in signals:
        print(f"{company_id}: {signal:.4f}")
